crates/valence_chat/src/lib.rs

```diff
@@ -0,0 +1,632 @@
+            >= session.session_data.expires_at as u128
+        {
+            warn!("Failed to validate profile key: expired public key");
+            commands.add(DisconnectClient {
+                client: session.client,
+                reason: Text::translate(MULTIPLAYER_DISCONNECT_EXPIRED_PUBLIC_KEY, []),
+            });
+            continue;
+        }
+
+        // Serialize the session data.
+        let mut serialized = Vec::with_capacity(318);
+        serialized.extend_from_slice(uuid.0.into_bytes().as_slice());
+        serialized.extend_from_slice(session.session_data.expires_at.to_be_bytes().as_ref());
+        serialized.extend_from_slice(session.session_data.public_key_data.as_ref());
+
+        // Hash the session data using the SHA-1 algorithm.
+        let mut hasher = Sha1::new();
+        hasher.update(&serialized);
+        let hash = hasher.finalize();
+
+        // Verify the session data using Mojang's public key and the hashed session data
+        // against the message signature.
+        if services_state
+            .public_key
+            .verify(
+                PaddingScheme::new_pkcs1v15_sign::<Sha1>(),
+                &hash,
+                session.session_data.key_signature.as_ref(),
+            )
+            .is_err()
+        {
+            warn!("Failed to validate profile key: invalid public key signature");
+            commands.add(DisconnectClient {
+                client: session.client,
+                reason: Text::translate(MULTIPLAYER_DISCONNECT_INVALID_PUBLIC_KEY_SIGNATURE, []),
+            });
+        }
+
+        // Decode the player's session public key from the data.
+        if let Ok(public_key) =
+            RsaPublicKey::from_public_key_der(session.session_data.public_key_data.as_ref())
+        {
+            // Update the player's chat state data with the new player session data.
+            state.chain.link = Some(MessageLink {
+                index: 0,
+                sender: uuid.0,
+                session_id: session.session_data.session_id,
+            });
+
+            // Add the chat session data to player.
+            // The player list will then send this new session data to the other clients.
+            commands.entity(session.client).insert(ChatSession {
+                public_key,
+                session_data: session.session_data.clone(),
+            });
+        } else {
+            // This shouldn't happen considering that it is highly unlikely that Mojang
+            // would provide the client with a malformed key. By this point the
+            // key signature has been verified.
+            warn!("Received malformed profile key data from '{}'", username.0);
+            commands.add(DisconnectClient {
+                client: session.client,
+                reason: Text::translate(
+                    MULTIPLAYER_DISCONNECT_EXPIRED_PUBLIC_KEY,
+                    ["Malformed profile key data".color(Color::RED)],
```

Comment:
Vanilla throws a DecoderException because the key is deserialized when the packet is decoded. This results in a disconnect with the `DISCONNECT_GENERIC` key and some text that looks something like "Internal Exception: Malformed public key bytes." Meaning it is a hardcoded English string.

![vanilla](https://user-images.githubusercontent.com/60993440/236729810-deea8ca1-6985-45de-9029-36f1a056815e.png)